### PR TITLE
RFC: Quantized Variables

### PR DESCRIPTION
**Pull request by lgeiger**
This is an initial implementation of #306 to explore the idea further.

## TODO
- [x] Figure out how to deal with variable assignment: https://github.com/tensorflow/tensorflow/issues/34332, https://github.com/tensorflow/tensorflow/pull/34779, fixed in f44f24eec8785015360db2043467c6cf19a1ed83
- [x] Add unit tests for `QuantizedVariable` and manually check special cases
  - [x] Test operator overloads (e.g. `__add__`, `__mul__`)
  - [x] Test Multi GPU training with `tf.DistributedVariable`
  - [x] Test Keras Float 16 training using `AutoCastVariable`
  - [x] Test Model checkpointing and saving, should be working correctly now
  - [x] Test Gradient computation in eager, functional and legacy graph mode
  - [x] Fix unit tests for TF 1.14
  - [x] Train CIFAR10 model to verify end-to-end workflow
    - [x] TensorFlow 2.0
    - [x] TensorFlow 1.14

### CODE REVIEW

File: larq/models.py [hunk 1]
```diff
@@ -230,14 +230,9 @@ def generate_table_row(self, table_config):
 
     def _get_bitwidth(self, weight):
         try:
-            for quantizer, quantized_weight in zip(
-                self._layer.quantizers, self._layer.quantized_latent_weights
-            ):
-                if quantized_weight is weight:
-                    return quantizer.precision
+            return weight.precision
         except:
```
Comment (leonoverweel) [inline larq/models.py]:
```suggestion
         except AttributeError:
```

Assuming this'd be an `AttributeError`, I think it's better to catch it more explicitly.

Comment (leonoverweel) [reply larq/models.py]: Or maybe:

```python 
    def _get_bitwidth(self, weight):
        return getattr(weight.precision, 32)
```

File: larq/optimizers_test.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import pytest
 import numpy as np
-import tensorflow as tf
```
Comment (leonoverweel) [inline larq/optimizers_test.py]: Why this change? In most of the codebase we make calls to `tf.keras.something`, not `keras.something`, right?

File: larq/layers_base.py [hunk 3]
```diff
@@ -1,14 +1,46 @@
...
-class QuantizerBase(tf.keras.layers.Layer):
+class BaseLayer(tf.keras.layers.Layer):
+    """Base class for defining quantized layers"""
+
+    def get_quantizer(self, name):
+        return None
+
+    def add_weight(self, name=None, **kwargs):
+        quantizer = self.get_quantizer(name)
+        if quantizer is not None:
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/layers_base.py]: Yes, we should revisit this function and see if we really need to rely on the private `base_layer_utils.make_variable` method. This is mainly copy paste from TensorFlow code.

File: larq/layers_base.py [hunk 4]
```diff
@@ -1,14 +1,46 @@
 import logging
 import tensorflow as tf
-from larq import quantizers, utils, metrics as lq_metrics
+from larq import quantizers, quantized_variable, metrics as lq_metrics
+from tensorflow.python.keras.engine import base_layer_utils
 
 log = logging.getLogger(__name__)
 
 
 # TODO: find a good way remove duplication between QuantizerBase, QuantizerDepthwiseBase and QuantizerSeparableBase
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/layers_base.py]: We're not quite there due to the metrics, but `BaseLayer` is a start.

File: larq/optimizers_test.py (hunk 2 above)
Comment (lgeiger) [reply larq/optimizers_test.py]: > Why this change?

No reason, mainly caused by some careless early morning debugging. 👍 Reverted the change

File: larq/layers_base.py [hunk 5]
```diff
@@ -34,35 +64,25 @@ def __init__(
                 "may result in starved weights (where the gradient is always zero)."
             )
 
+    def get_quantizer(self, name):
+        if name == "kernel":
+            return self.kernel_quantizer
+        return None
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/layers_base.py]: Me too

File: larq/models.py (hunk 1 above)
Comment (lgeiger) [reply larq/models.py]: Fixed in 2f79a608529764e6c7bd944deeb44aa4268155c7

File: larq/layers_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (leonoverweel) [reply larq/layers_base.py]: @lgeiger I probably shouldn't have put this as suggested change because the 

```python
        if name == "kernel":
            return self.kernel_quantizer
```

is still there now.

File: larq/quantized_variable.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
+"""Contains QuantizedVariable, a variable that can be quantized in the forward pass."""
+from tensorflow.python.distribute import values as distribute_values
+from tensorflow.python.eager import context
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import resource_variable_ops
+from tensorflow.python.ops import variables
+from functools import wraps
+
+
+def quantize(method):
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/quantized_variable.py]: We could also do it the same way we deal with operators and and keep a list of methods we want to delegate to the `latent_variable` and methods where we want to return the quantized version and then use `setattr` to set the method on the class.

Though I think the current way is a lot simpler, although it results in a lot more code.

File: larq/layers_base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/layers_base.py]: Done in 3ce44e88ffcf4ef6f49a22dc382022f5e8e7c4cf

**Comment by lgeiger** [comment]
I think this is ready to take for a spin. I added a TODO list above for the things that still need to be done before merging.

File: larq/quantized_variable.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgeiger) [reply larq/quantized_variable.py]: I had to remove the nice decorator in fe195dd5a26cff390dba04bbfb4fc3d3125a9555 since I forgot about some other cases we need to handle.

**Comment by lgeiger** [comment]
This is ready for a first round of reviews. I'll add more unittesting on the flight to NeurIPS.

File: larq/conftest.py [hunk 7]
```diff
@@ -14,14 +14,20 @@ def eager_mode():
 def graph_mode():
     """pytest fixture for running test in graph mode"""
     with context.graph_mode():
-        yield
+        with tf.compat.v1.Session().as_default():
```
Comment (leonoverweel) [inline larq/conftest.py]: What is this for?

File: larq/conftest.py [hunk 8]
```diff
@@ -46,3 +52,12 @@ def keras_should_run_eagerly(request):
     else:
         with context.eager_mode():
             yield request.param == "tf_keras_eager"
+
+
+@pytest.fixture(params=[False, True])
```
Comment (leonoverweel) [inline larq/conftest.py]: It might be nicer to make these params strings describing what the different settings are instead of booleans, mirroring previous fixtures:
- `eager_and_graph_mode` (`["eager", "graph"]`) and
- `keras_should_run_eagerly ` (`["graph", "tf_eager", "tf_keras_eager"]`)

File: larq/layers_base.py [hunk 9]
```diff
@@ -3,15 +3,47 @@
...
+        quantizer = self.get_quantizer(name)
+        if quantizer is None:
+            return super()._add_variable_with_custom_getter(name, **kwargs)
+
+        old_getter = kwargs.pop("getter")
+
+        # Wrap `getter` with a version that returns a `QuantizedVariable`.
+        def getter(*args, **kwargs):
+            variable = old_getter(*args, **kwargs)
+            return quantized_variable.create_quantized_variable(variable, quantizer)
```
Comment (leonoverweel) [inline larq/layers_base.py]: What do you think of doing this as a [factory pattern](https://pythonspot.com/factory-method/)? The call would then be:

```python
return quantized_variable.QuantizedVariable.factory(variable, quantizer)
```

This may just be my Java background talking though 😅

File: larq/quantized_variable.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+        # Arguments
+        variable: A floating-point resource variable to wrap.
+        quantizer: An optional quantizer to transform the floating-point
+            variable to a fake quantized variable.
+        precision: An optional integer defining the precision of the quantized
+            variable. If `None`, `quantizer.precision` is used.
+        """
+        if not resource_variable_ops.is_resource_variable(variable):
+            raise ValueError(
+                f"variable must be of type tf.ResourceVariable, but got: {variable}"
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: Do we want to report `type(variable)` here?

File: larq/quantized_variable.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+            )
+        if not (quantizer is None or callable(quantizer)):
+            raise ValueError(
+                f"quantizer must be callable or None, but got: {quantizer}"
+            )
+        if not (precision is None or type(precision) == int):
+            raise ValueError(
+                f"precision must be of type integer or None, but got: {precision}"
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: Also report `type(precision)` here?

File: larq/quantized_variable.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        precision: An optional integer defining the precision of the quantized
+            variable. If `None`, `quantizer.precision` is used.
+        """
+        if not resource_variable_ops.is_resource_variable(variable):
+            raise ValueError(
+                f"variable must be of type tf.ResourceVariable, but got: {variable}"
+            )
+        if not (quantizer is None or callable(quantizer)):
+            raise ValueError(
+                f"quantizer must be callable or None, but got: {quantizer}"
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: Also report `type(quantizer)` here?

File: larq/quantized_variable.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        return self._quantize(self.latent_variable).numpy()
+
+    def sparse_read(self, *args, **kwargs):
+        return self._quantize(self.latent_variable.sparse_read(*args, **kwargs))
+
+    def gather_nd(self, *args, **kwargs):
+        return self._quantize(self.latent_variable.gather_nd(*args, **kwargs))
+
+    def __getattr__(self, name):
+        return getattr(self.latent_variable, name)
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: Following [this SO post](https://stackoverflow.com/a/1944708/1814949), should we do error handling for nonexistent names here?

File: larq/quantized_variable.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+
+QuantizedVariable._OverloadAllOperators()
+tf.register_tensor_conversion_function(
+    QuantizedVariable, QuantizedVariable._dense_var_to_tensor
+)
+ops.register_dense_tensor_like_type(QuantizedVariable)
+
+
+def create_quantized_variable(variable, quantizer=None, precision=None):
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: See above; I think this would make sense as a factory within the `QuantizedVariable` class.

File: larq/quantized_variable.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    if wrap and resource_variable_ops.is_resource_variable(variable):
+        return create_quantized_variable(variable, quantizer, precision)
+    return variable
+
+
+def _get_name(obj):
+    try:
+        return obj.__name__
+    except AttributeError:
+        return obj.__class__.__name__
```
Comment (leonoverweel) [inline larq/quantized_variable.py]: Since this is just used in `QuantizedVariable`, I think it may be better to make it a static internal method of that class.

File: larq/quantized_variable_test.py [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+
+def get_var(val, dtype=None, name=None):
+    return tf.compat.v1.Variable(val, use_resource=True, dtype=dtype, name=name)
+
+
+@pytest.fixture(params=[True, False])
+def quantized(request):
+    """pytest fixture for running test quantized and non-quantized"""
+    with quantized_scope.scope(request.param):
+        yield request.param
```
Comment (leonoverweel) [inline larq/quantized_variable_test.py]: I think using strings instead of bools would be nicer here too.

Comment (leonoverweel) [reply larq/quantized_variable_test.py]: And maybe we can move this into `conftest.py`, where most other pytest fixtures are. (There's only one other fixture that's not in that file: one in `models_test.py`; I think moving them all to `conftest.py` to have one place to look for fixtures would make sense.)

File: larq/conftest.py (hunk 7 above)
Comment (lgeiger) [reply larq/conftest.py]: This is so that [`evaluate`](https://github.com/larq/larq/pull/312/files#diff-915b33c29d82e4e01202fe2566e0f20fR26) can evaluate the tensor in a shared session per unittest.

File: larq/conftest.py (hunk 8 above)
Comment (lgeiger) [reply larq/conftest.py]: I opted for booleans so it is easy to check whether we are running the unittest distributed (see [here](https://github.com/larq/larq/pull/312/files#diff-2c6d294a8f4ae0b3b2ee3df2966dca03R31))

File: larq/layers_base.py (hunk 9 above)
Comment (lgeiger) [reply larq/layers_base.py]: Hmm I opted for this since it is the way `AutoCastVariable` handles it is pretty simple. I'd be fine with something like `QuantizedVariable.from_variable` though. Since we currently don't expose the `QuantizedVariable` to the user by default, we can also change this in the future.

File: larq/quantized_variable.py (hunk 10 above)
Comment (lgeiger) [reply larq/quantized_variable.py]: Yeah might be a good thing. I only copied this input validation from `AutoCastVariable` without thinking to much about it ;)

File: larq/quantized_variable.py (hunk 13 above)
Comment (lgeiger) [reply larq/quantized_variable.py]: Not sure. I think it's fine not to handle any errors here, since `latent_variable` variable will throw an readable error message  if the attribute doesn't exist.

File: larq/quantized_variable.py (hunk 14 above)
Comment (lgeiger) [reply larq/quantized_variable.py]: How do you feel about violating PEP8 and naming this function `QuantizedVariable` so it feels like a class when using it?

File: larq/quantized_variable.py (hunk 15 above)
Comment (lgeiger) [reply larq/quantized_variable.py]: :+1: Makes sense.

File: larq/quantized_variable_test.py (hunk 16 above)
Comment (lgeiger) [reply larq/quantized_variable_test.py]: I think I like returning booleans since it makes using the fixture in the test very easy since I only have to do `if quantized:` and not check if the return value matches a string. I put it here since this is the only file we need to use it.

File: larq/conftest.py (hunk 8 above)
Comment (leonoverweel) [reply larq/conftest.py]: OK makes sense.

File: larq/layers_base.py (hunk 9 above)
Comment (leonoverweel) [reply larq/layers_base.py]: OK I'm fine with that as well.

File: larq/quantized_variable.py (hunk 14 above)
Comment (leonoverweel) [reply larq/quantized_variable.py]: It feels a bit wrong but I think it might make sense here.

File: larq/quantized_variable_test.py (hunk 16 above)
Comment (leonoverweel) [reply larq/quantized_variable_test.py]: OK booleans are good with me.

For the fixture location: we also have the `distribute_scope` and `eager_and_graph_mode` fixtures in `conftest.py` even though they only get used by `quantized_variable_test.py`, so I think we should treat the `quantized` fixture the same. We could then also test it in `conftest_test.py` like the other two.

File: larq/quantized_variable.py (hunk 14 above)
Comment (lgeiger) [reply larq/quantized_variable.py]: > It feels a bit wrong but I think it might make sense here.

Yeah I also feel this would be kinda wrong
